queue/group_remote.go

```diff
@@ -0,0 +1,387 @@
+	}
+	defer c.Close(ctx)
+	var collections []string
+	for c.Next(ctx) {
+		elem := listCollectionsOutput{}
+		if err := c.Decode(&elem); err != nil {
+			return nil, errors.Wrap(err, "problem parsing listCollections output")
+		}
+		collections = append(collections, elem.Name)
+	}
+	if c.Err() != nil {
+		return nil, errors.Wrap(err, "problem iterating over list collections cursor")
+	}
+	return collections, nil
+}
+
+// Get a queue with the given index. Get sets the last accessed time to now. Note that this means
+// that the caller must add a job to the queue within the TTL, or else it may have attempted to add
+// a job to a closed queue.
+func (g *remoteQueueGroup) Get(ctx context.Context, id string) (amboy.Queue, error) {
+	g.mu.RLock()
+	if queue, ok := g.queues[id]; ok {
+		g.ttlMap[id] = time.Now()
+		g.mu.RUnlock()
+		return queue, nil
+	}
+	g.mu.RUnlock()
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	// Check again in case the map was modified after we released the read lock.
+	if queue, ok := g.queues[id]; ok {
+		g.ttlMap[id] = time.Now()
+		return queue, nil
+	}
+
+	queue, err := g.startProcessingRemoteQueue(ctx, collectionFromID(g.prefix, id))
+	if err != nil {
+		return nil, errors.Wrap(err, "problem starting queue")
+	}
+	g.queues[id] = queue
+	g.ttlMap[id] = time.Now()
+	return queue, nil
+}
+
+// Put a queue at the given index. The caller is responsible for starting thq queue.
+func (g *remoteQueueGroup) Put(ctx context.Context, id string, queue amboy.Queue) error {
+	g.mu.RLock()
+	if _, ok := g.queues[id]; ok {
+		g.mu.RUnlock()
+		return errors.New("a queue already exists at this index")
+	}
+	g.mu.RUnlock()
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	// Check again in case the map was modified after we released the read lock.
+	if _, ok := g.queues[id]; ok {
+		return errors.New("a queue already exists at this index")
+	}
+
+	g.queues[id] = queue
+	g.ttlMap[id] = time.Now()
+	return nil
+}
+
+// Prune queues that have no pending work, and have completed work older than the TTL.
+func (g *remoteQueueGroup) Prune(ctx context.Context) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	colls, err := getExistingCollections(ctx, g.client, g.mongooptions.DB, g.prefix)
+	if err != nil {
+		return errors.Wrap(err, "problem getting collections")
+	}
+	for i, coll := range colls {
+		// This is an optimization. If we've added to the queue recently enough, there's no
+		// need to query its contents, since it cannot be old enough to prune.
+		if t, ok := g.ttlMap[coll]; ok && time.Since(t) < g.ttl {
+			remove(colls, i)
+		}
+	}
+	catcher := grip.NewBasicCatcher()
+	wg := &sync.WaitGroup{}
+	collsDeleteChan := make(chan string, len(colls))
+	collsDropChan := make(chan string)
+
+	go func() {
+		for _, coll := range colls {
+			collsDropChan <- coll
+		}
+		close(collsDropChan)
+	}()
+	wg = &sync.WaitGroup{}
+	for i := 0; i < runtime.NumCPU(); i++ {
+		wg.Add(1)
+		go func(ch chan string) {
```

Comment:
I think you don't need to pass the channel as a function.

Reply:
no, but I think it's nice to be in the habit, since there are gotchas if you don't pass args to goroutines